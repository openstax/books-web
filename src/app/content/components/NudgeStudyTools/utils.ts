import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides, studyGuidesEnabled } from '../../studyGuides/selectors';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudge,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean, windowWidth: number): Positions => {
  const { top, left, right, height, width } = getBoundingRectOfNudgeTarget(target);
  const padding = remsToPx(spotlightPadding);
  const spotlightTopOffset = top - padding;
  const spotlightLeftOffset = left - padding;
  const spotlightHeight = height + (padding * 2);
  const spotlightWidth = width + (padding * 2);
  const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
  const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperRight = windowWidth - right - padding;
  const closeButtonLeft = left + spotlightWidth;
  const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperRight,
    contentWrapperTopOffset,
    spotlightHeight,
    spotlightLeftOffset,
    spotlightTopOffset,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  // Starting values depends on if we checks for Math.max or Math.min
  let top = 9999;
  let bottom = 0;
  let left = 9999;
  let right = 0;

  for (const child of Array.from(target.children)) {
    const rect = child.getBoundingClientRect();
    top = Math.min(top, rect.top);
    bottom = Math.max(bottom, rect.bottom);
    left = Math.min(left, rect.left);
    right = Math.max(right, rect.right);
  }

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);
  const isEnabled = useSelector(studyGuidesEnabled);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null;
    if (isEnabled && target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results with different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile, windowWidth));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [studyGuides, isEnabled, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudge.pageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString());
  Cookies.set(cookieNudge.date, now.toString());
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
