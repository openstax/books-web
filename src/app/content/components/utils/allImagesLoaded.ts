import { Document,  Element } from '@openstax/types/lib.dom';

export default (container: Element | Document) => {
  return Promise.all(Array.from(container.querySelectorAll('img')).map((img) => img.complete
    ? Promise.resolve()
    : new Promise<void>((resolve) => {

      const previousOnload = img.onload;
      const previousOnerror = img.onerror;

      img.onload = (...args) => {
        resolve();
        if (previousOnload) {
          previousOnload.apply(img, args);
        }
      };
      img.onerror = (...args) => {
        resolve();
        if (previousOnerror) {
          previousOnerror.apply(img, args);
        }
      };
    })
  )).then((): void => undefined);
};
