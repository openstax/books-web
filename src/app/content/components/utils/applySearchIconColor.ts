import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { SearchButtonColor } from '../../types';
import { toolbarIconColor } from '../constants';

export const applySearchIconColor = (props: {colorSchema: SearchButtonColor}) =>
  props.colorSchema ? css`
    color: ${theme.color.primary[props.colorSchema].foreground};

    :hover,
    :focus {
        color: ${theme.color.primary[props.colorSchema].foreground};

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
