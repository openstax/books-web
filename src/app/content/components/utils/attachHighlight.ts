import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
//import Sentry from '../../../../helpers/Sentry';

const attachHighlight = (
  highlight: Highlight | SerializedHighlight,
  highlighter: Highlighter
) => {
  highlighter.highlight(highlight);
  const result = highlighter.getHighlight(highlight.id);
  if (!result || !result.isAttached()) {
    console.log(`Highlight with id: ${highlight.id} has not been attached.`)
    // Sentry.captureException(new Error(`Highlight with id: ${highlight.id} has not been attached.`));
  }
  return result;
};

export default attachHighlight;
