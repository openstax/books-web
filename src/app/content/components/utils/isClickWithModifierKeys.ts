import { MouseEvent } from '@openstax/types/lib.dom';

export default  (e: React.MouseEvent | MouseEvent) => e.shiftKey || e.ctrlKey || e.metaKey || e.altKey;
