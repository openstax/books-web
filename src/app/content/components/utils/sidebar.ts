import { connect } from 'react-redux';
import { css, FlattenSimpleInterpolation } from 'styled-components';
import theme from '../../../theme';
import { AppState } from '../../../types';
import * as selectors from '../../selectors';
import { State } from '../../types';

export const isOpenConnector = connect((state: AppState) => ({isOpen: selectors.tocOpen(state)}));

export const styleWhenSidebarClosed = (closedStyle: FlattenSimpleInterpolation) => css`
  ${(props: {isOpen: State['tocOpen']}) => props.isOpen === null && theme.breakpoints.mobile(closedStyle)}
  ${(props: {isOpen: State['tocOpen']}) => props.isOpen === false && closedStyle}
`;
