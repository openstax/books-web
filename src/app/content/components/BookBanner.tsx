import React, { Component } from 'react';
import { connect } from 'react-redux';
import withServices from '../../context/Services';
import styled from 'styled-components';
import theme from '../../theme';
import { AppServices, AppState } from '../../types';
import * as select from '../selectors';
import { Book, Page } from '../types';
import {ChevronLeft} from 'styled-icons/boxicons-regular/ChevronLeft'
import { findArchiveTreeSection, bookDetailsUrl } from '../utils'
import Color from 'color';
import typography from '../../components/Typography';

const LeftArrow = styled(ChevronLeft)`
  height: 2rem;
  width: 2rem;
  color: #FFFFFF;
  margin-right: 0.7rem;
`;

interface PropTypes {
  page?: Page;
  book?: Book;
  services: AppServices;
}

const TopBar = styled.div`
  width: 100%;
  max-width: 117rem;
  background: transparent;
  margin: 0 auto;
  text-align: left;
`;

const BookTitle = styled.a`
  ${typography.h4Style}
  letter-spacing: -0.04rem;
  color: ${theme.color.primary.blue.foreground};
  text-align: left;
  display: inline-block;
  width: 100%;
  max-width: 87rem;
  margin: 0;
  text-decoration: none;
`;

const BookChapter = styled.h3`
  ${typography.h3Style}
  letter-spacing: -0.04rem;
  color: ${theme.color.primary.blue.foreground};
  font-weight: bold;
  text-align: left;
  display: inline-block;
  width: 100%;
  max-width: 87rem;
  margin-top: 1rem;
  margin-bottom: 0;
`;

const blue = `${theme.color.primary.blue.base}`;
const color = Color(`${blue}`).lighten(0.7); 

const BarWrapper = styled.div`
  width: 100%;
  padding: ${theme.contentBuffer.default.padding};
  box-shadow: 0 0.2rem 0.2rem 0 rgba(0, 0, 0, 0.1);
  display: flex;
  align-items: center;
  height: 13rem;
  background: linear-gradient(to right, ${blue}, ${color.hex()});
`;

export class TitleComponent extends Component<PropTypes> {
  public render() {
    const {page, book} = this.props as PropTypes;
    
    if (!book || !page) {
      return null
    }

    const treeSection = findArchiveTreeSection(book, page.id);
    const bookUrl = bookDetailsUrl(book);

    if(!treeSection) {
      return null;
    }

    return <BarWrapper>
      <TopBar>
        <BookTitle href={bookUrl}><LeftArrow/>{book.tree.title}</BookTitle>
        <BookChapter dangerouslySetInnerHTML={{__html: treeSection.title}}></BookChapter>
      </TopBar>
    </BarWrapper>;
  }
}

export default connect(
  (state: AppState) => ({
    book: select.book(state),
    page: select.page(state),
  })
)(withServices(TitleComponent));

