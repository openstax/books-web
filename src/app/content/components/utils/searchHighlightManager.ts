import Highlighter from '@openstax/highlighter';
import { SearchResultHit } from '@openstax/open-search-client';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import isEqual from 'lodash/fp/isEqual';
import { scrollTo } from '../../../utils';
import { SelectedResult } from '../../search/types';
import { highlightResults } from '../../search/utils';
import allImagesLoaded from './allImagesLoaded';

interface Services {
  highlighter: Highlighter;
  container: HTMLElement;
  searchResults?: SearchResultHit[];
  searchResultMap: ReturnType<typeof highlightResults>;
  selected?: SelectedResult | null;
}

const scrollToSearch = ({container, searchResultMap}: Services, selected: SelectedResult) => {
  const elementHighlights = searchResultMap.find((map) => isEqual(map.result, selected.result));
  const selectedHighlights = elementHighlights && elementHighlights.highlights[selected.highlight];
  const firstSelectedHighlight = selectedHighlights && selectedHighlights[0];

  if (firstSelectedHighlight) {
    firstSelectedHighlight.focus();
    allImagesLoaded(container).then(
      () => scrollTo(firstSelectedHighlight.elements[0])
    );
  }
};

const updateResults = (services: Services, searchResults: SearchResultHit[]): Services => {
  if (services.searchResults === searchResults) {
    return services;
  }

  services.highlighter.eraseAll();

  const searchResultMap = highlightResults(services.highlighter, searchResults);

  return {...services, searchResultMap};
};

const selectResult = (services: Services, selected: SelectedResult | null): Services => {
  services.highlighter.clearFocus();

  if (selected && services.selected !== selected) {
    scrollToSearch(services, selected);
  }

  return {...services, selected};
};

const searchHighlightManager = (initialServices: Services) =>
  (searchResults: SearchResultHit[], selected: SelectedResult | null) => {
    const newServices = flow(
      (services) => updateResults(services, searchResults),
      (services) => selectResult(services, selected)
    )(initialServices);

    return searchHighlightManager(newServices);
  };

export default (container: HTMLElement) => {
  const services = {
    container,
    highlighter: new Highlighter(container, {
      className: 'search-highlight',
    }),
    searchResultMap: [],
  };

  return searchHighlightManager(services);
};

export const stubManager: ReturnType<typeof searchHighlightManager> = () => stubManager;
