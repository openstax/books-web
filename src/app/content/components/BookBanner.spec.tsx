import React from 'react';
import renderer from 'react-test-renderer';
import { book as archiveBook, shortPage } from '../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../test/mocks/osWebLoader';
import { renderToDom } from '../../../test/reactutils';
import { resetModules } from '../../../test/utils';
import { assertWindow } from '../../utils';
import { formatBookData } from '../utils';
import { findArchiveTreeNodeById } from '../utils/archiveTreeUtils';
import { BarWrapper, PropTypes } from './BookBanner';

const book = formatBookData(archiveBook, mockCmsBook);
const bookWithoutOsWebData = formatBookData(archiveBook, undefined);
const pageNode = findArchiveTreeNodeById(archiveBook.tree, shortPage.id)!;

describe('BookBanner', () => {
  let window: Window;
  let event: React.MouseEvent;
  let assign: jest.SpyInstance;
  // tslint:disable-next-line:variable-name
  let BookBanner: React.ComponentType<PropTypes>;

  beforeEach(() => {
    resetModules();

    window = assertWindow();
    delete window.location;

    window.location = {
      assign: jest.fn(),
    } as any as Window['location'];

    event = {
      altKey: false,
      ctrlKey: false,
      metaKey: false,
      preventDefault: jest.fn(),
      shiftKey: false,
    } as any as React.MouseEvent;

    assign = jest.spyOn(window.location, 'assign');
  });

  describe('without unsaved changes', () => {
    beforeEach(() => {
      BookBanner = require('./BookBanner').BookBanner;
    });

    it('renders empty state with no page or book', () => {
      const component = renderer.create(<BookBanner />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly when you pass a page and book', () => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders correctly without osweb data', () => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={bookWithoutOsWebData} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('does not stop default navigation event', async() => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} hasUnsavedHighlight={false} />);

      const link = component.root.findByProps({'data-testid': 'details-link-expanded'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).not.toHaveBeenCalled();
    });

    it('mounts in a dom', () => {
      expect(() => renderToDom(<BookBanner pageNode={pageNode} book={book} />)).not.toThrow();
    });

    it('wrapper transition matches snapshot', () => {
      const component = renderer.create(<BarWrapper colorSchema='blue' up={true} />);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('with unsaved changes', () => {
    const mockConfirmation = jest.fn()
      .mockImplementationOnce(() =>  new Promise((resolve) => resolve(true)))
      .mockImplementationOnce(() =>  new Promise((resolve) => resolve(false)));

    jest.mock(
      '../highlights/components/utils/showConfirmation',
      () => mockConfirmation
    );

    beforeEach(() => {
      BookBanner = require('./BookBanner').BookBanner;
    });

    it('redirects if users chooses to discard', async() => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} hasUnsavedHighlight={true} />);

      const link = component.root.findByProps({'data-testid': 'details-link-expanded'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).toHaveBeenCalled();
      expect(assign).toHaveBeenCalled();
    });

    it('noops if users chooses not to discard', async() => {
      const component = renderer.create(<BookBanner pageNode={pageNode} book={book} hasUnsavedHighlight={true} />);

      const link = component.root.findByProps({'data-testid': 'details-link-collapsed'});

      await renderer.act(() => {
        link.props.onClick(event);
        return Promise.resolve();
      });

      expect(event.preventDefault).toHaveBeenCalled();
      expect(assign).not.toHaveBeenCalled();
    });
  });
});
