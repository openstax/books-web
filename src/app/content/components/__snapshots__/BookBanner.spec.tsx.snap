// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookBanner without unsaved changes renders correctly when you pass a page and book 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-4 eIPJBY"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitle-sc-1avy0c0-2 fRzrrt"
        data-testid="details-link-expanded"
        href="/details/books/book-slug-1"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 eRQNag StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <h1
        className="BookBanner__BookChapter-sc-1avy0c0-3 dMXgkT"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-4 gZpRfI"
    data-analytics-region="book-banner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitle-sc-1avy0c0-2 kplCmc"
        data-testid="details-link-collapsed"
        href="/details/books/book-slug-1"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 eRQNag StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <span
        className="BookBanner__BookChapter-sc-1avy0c0-3 fVxHGG"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders correctly without osweb data 1`] = `
Array [
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-4 eIPJBY"
    data-analytics-region="book-banner-expanded"
    data-testid="bookbanner"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitle-sc-1avy0c0-2 fRzrrt"
        data-testid="details-link-expanded"
        href="/errors/404"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 eRQNag StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <h1
        className="BookBanner__BookChapter-sc-1avy0c0-3 dMXgkT"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
  <div
    className="BookBanner__BarWrapper-sc-1avy0c0-4 gZpRfI"
    data-analytics-region="book-banner-collapsed"
  >
    <div
      className="BookBanner__TopBar-sc-1avy0c0-1 jVswHX"
    >
      <a
        className="BookBanner__BookTitle-sc-1avy0c0-2 kplCmc"
        data-testid="details-link-collapsed"
        href="/errors/404"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="BookBanner__LeftArrow-sc-1avy0c0-0 eRQNag StyledIconBase-sc-bdy9j4 jbPeIR"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13.293 6.293L7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
          />
        </svg>
        Test Book 1
      </a>
      <span
        className="BookBanner__BookChapter-sc-1avy0c0-3 fVxHGG"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"os-text\\">Test Page 4</span>",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`BookBanner without unsaved changes renders empty state with no page or book 1`] = `
<div
  className="BookBanner__BarWrapper-sc-1avy0c0-4 kLMCbe"
/>
`;

exports[`BookBanner without unsaved changes wrapper transition matches snapshot 1`] = `
<div
  className="BookBanner__BarWrapper-sc-1avy0c0-4 axkrA"
/>
`;
